google/oauth2/service_account.py

```diff
@@ -336,3 +336,207 @@ def signer(self):
     @_helpers.copy_docstring(credentials.Signing)
     def signer_email(self):
         return self._service_account_email
+
+
+class IDTokenCredentials(credentials.Signing, credentials.Credentials):
+    """Open ID Connect ID Token-based service account credentials.
+
+    These credentials are largely similar to :class:`.Credentials`, but instead
+    of using an OAuth 2.0 Access Token as the bearer token, they use an Open
+    ID Connect ID Token as the bearer token. These credentials are useful when
+    communicating to services that require ID Tokens and can not accept access
+    tokens.
+
+    Usually, you'll create these credentials with one of the helper
+    constructors. To create credentials using a Google service account
+    private key JSON file::
+
+        credentials = (
+            service_account.IDTokenCredentials.from_service_account_file(
+                'service-account.json'))
+
+    Or if you already have the service account file loaded::
+
+        service_account_info = json.load(open('service_account.json'))
+        credentials = (
+            service_account.IDTokenCredentials.from_service_account_info(
+                service_account_info))
+
+    Both helper methods pass on arguments to the constructor, so you can
+    specify additional scopes and a subject if necessary::
+
+        credentials = (
+            service_account.IDTokenCredentials.from_service_account_file(
+                'service-account.json',
+                scopes=['email'],
+                subject='user@example.com'))
+`
+    The credentials are considered immutable. If you want to modify the scopes
+    or the subject used for delegation, use :meth:`with_scopes` or
+    :meth:`with_subject`::
+
+        scoped_credentials = credentials.with_scopes(['email'])
+        delegated_credentials = credentials.with_subject(subject)
+
+    """
+    def __init__(self, signer, service_account_email, token_uri,
```

Comment:
It seems like this constructor (and some of the factories / factory helpers below) is **very** similar to existing code. Is this a problem @jonparrott?

Reply:
I really struggled with this. I considered quite a few options here:

* making a common class for `service_account.Credentials` and `service_account.IDTokenCredentials`.
* making `IDTokenCredentials` a sub-class of `service_account.Credentials`, but it doesn't work because `IDTokenCredentials are actually scopeless.
* making `IDTokenCredentials` a sub-class of `jwt.Credentials` (as in the original PR, but it doesn't work because certain methods don't make sense and there was one method (which I can't remember) that would break the substitution principle - basically, `IDTokenCredentials` could not reasonably ever take the place of `jwt.Credentials`.
* Just make `IDTokenCredentials` its own class with similar methods.

The only viable options were 1 and 4, and I ended up with four, but I'm happy to reconsider (1) or some alternative I didn't think of.

Reply:
I'm going to merge this as-is. If this becomes onerous we can refactor out without breaking the interface.